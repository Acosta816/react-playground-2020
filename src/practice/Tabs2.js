import React from 'react';


class Tabs2 extends React.Component {

    static defaultProps = {
        tabs: []
    }

    state = {
        currentIndex: [0]
    }

    renderButtons = ()=> {
        const myButtons = this.props.tabs.map((tab,index)=> {
            return (
                <button onClick={()=> {this.handleClick(index)}} key={index}>{tab.name}</button>
            )
        })

        return myButtons;
    };

    renderContent = ()=> {
        const currentTab = this.props.tabs[this.state.currentIndex];
        return (
            <div>
                <h2>{currentTab.content}</h2>
                <img src={currentTab.avatar} alt="sonic chars"></img>
            </div>
            )
    }

    handleClick = (index)=> {
        console.log(index);
        this.setState({
            currentIndex: index
        })
    }

    render() {

        return(
            <div>
                {this.renderButtons()}
                {this.props.tabs.length ? this.renderContent(): null}
            </div>
        )
    }
}


export default Tabs2;