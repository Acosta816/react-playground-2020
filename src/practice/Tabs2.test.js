import React from 'react';
import ReactDOM from 'react-dom';
import Tabs2 from './Tabs2';
import renderer from 'react-test-renderer';// installed separately 
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json'; // installed separately 

/*
    ðŸ“· Snapshots  x3
*/

describe('TestSuite: Tabs2 Component', ()=> {
    //setup any variables here:
    const tabs = [{name:'Sonic', content: 'Fastest thing Alive', avatar: "https://b7.pngbarn.com/png/547/724/sonic-the-hedgehog-2-sonic-3d-sonic-mania-sonic-cd-sonic-the-hedgehog-png-clip-art.png"}, {name:'Tails', content: 'Loyal and Hyper-intelligent Flying Fox', avatar: "https://vignette.wikia.nocookie.net/crystal-crossover/images/c/cb/ManiaTailsFront.png/revision/latest?cb=20180805022514"}, {name:'Amy', content: 'Sonics Biggest Fan. Hammer-weilding Pink Hedgehog', avatar: 'https://b7.pngbarn.com/png/319/82/amy-rose-sonic-colors-sonic-chaos-sonic-cd-knuckles-the-echidna-sonic-the-hedgehog-png-clip-art.png'}, {name:'Robotnik', content: 'Crazed Genius, obsessed with robotisizing all life and ridding the world of Sonic'}, {name:'Shadow', content: 'Sonics adversary. Created by Robotnik to become the ultimate lifeform'}, {name:'Sally', content: 'Rightful Princess of Mobius and one of the Leaders in the Freedom Fighters against Robotnik. Sonics Best friend.'}];

    it('Smoke: renders without crashing', ()=> {
        const div = document.createElement('div');

        ReactDOM.render(<Tabs2 />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('ðŸ“·Snapshot: renders empty given no tabs props', ()=> {
        //wrap your component in an enzyme shallow wrapper where it's children wont be present.
        const enzWrapper = shallow(<Tabs2/>);
        //in order for Jest's Snapshots to work with the enzyme wrapper, we need to convert it to Json format using a method called toJson() from the 'enzyme-to-json' library installed separately .
        // like so... toJson(enzWrapper);
        // then we simply Jest snapshot test it like before and pass in Json(enzwrapper) instead of tree...
        expect(toJson(enzWrapper)).toMatchSnapshot();
    })


    it('ðŸ“·Snapshot: renders the first tab by default', ()=> {
        const enzWrapper = shallow(<Tabs2 tabs={tabs}/>);
        expect(toJson(enzWrapper)).toMatchSnapshot();
    })

    /********We are REPLACING react-test-renderer code with enzyme code since enzyme also renders when we create the shallow wrapper. The expect().toMatchSnapshot() is JEST btw so we will still use that code to create .snap files (snapshots). */
    
    // it('Snapshot: renders the first tab by default', ()=> {

    //     //create tree and mount the <Tabs/> with props in it.. <Tabs tabs={tabs}/>
    //     const tree = renderer.create(<Tabs2 tabs={tabs} />);

    //     //this statement is the test and it's also generating a snapshot with the last bit.
    //     expect(tree).toMatchSnapshot();

    // });

    it('ðŸ“·Snapshot: closes the first tab and opens ANY clicked tab', ()=> {
        //create the enzyme wrapper with props loaded.
        const enzWrapper = shallow(<Tabs2 tabs={tabs} />);
        //Next: traverse to one of the buttons to simulate a physical click...

        //This .debug() shows the HTML for the current wrapper's .find() selection.
        //  console.log(enzWrapper.find('button').debug() would show all buttons.
        //  console.log(enzWrapper.find('button').at(2).debug())  would show just the button at index#2 which is the 3rd button.
        console.log(enzWrapper.find('button').at(2).debug())

        //Traverse to the "3rd" button at index 2 and Simulate a Click!
        enzWrapper.find('button').at(2).simulate('click');

        //Make a Json version of the above simulation and pass it over to Jest's expect() testing so Jest can compare it to the snapshot it's about to take with toMatchSnapshot()ðŸ“·
        expect(toJson(enzWrapper)).toMatchSnapshot();


    })

})