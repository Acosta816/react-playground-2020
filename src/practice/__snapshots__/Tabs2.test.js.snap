// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestSuite: Tabs2 Component ðŸ“·Snapshot: closes the first tab and opens ANY clicked tab 1`] = `
<div>
  <button
    key="0"
    onClick={[Function]}
  >
    Sonic
  </button>
  <button
    key="1"
    onClick={[Function]}
  >
    Tails
  </button>
  <button
    key="2"
    onClick={[Function]}
  >
    Amy
  </button>
  <button
    key="3"
    onClick={[Function]}
  >
    Robotnik
  </button>
  <button
    key="4"
    onClick={[Function]}
  >
    Shadow
  </button>
  <button
    key="5"
    onClick={[Function]}
  >
    Sally
  </button>
  <div>
    <h2>
      Sonics Biggest Fan. Hammer-weilding Pink Hedgehog
    </h2>
    <img
      alt="sonic chars"
      src="https://b7.pngbarn.com/png/319/82/amy-rose-sonic-colors-sonic-chaos-sonic-cd-knuckles-the-echidna-sonic-the-hedgehog-png-clip-art.png"
    />
  </div>
</div>
`;

exports[`TestSuite: Tabs2 Component ðŸ“·Snapshot: renders empty given no tabs props 1`] = `<div />`;

exports[`TestSuite: Tabs2 Component ðŸ“·Snapshot: renders the first tab by default 1`] = `
<div>
  <button
    key="0"
    onClick={[Function]}
  >
    Sonic
  </button>
  <button
    key="1"
    onClick={[Function]}
  >
    Tails
  </button>
  <button
    key="2"
    onClick={[Function]}
  >
    Amy
  </button>
  <button
    key="3"
    onClick={[Function]}
  >
    Robotnik
  </button>
  <button
    key="4"
    onClick={[Function]}
  >
    Shadow
  </button>
  <button
    key="5"
    onClick={[Function]}
  >
    Sally
  </button>
  <div>
    <h2>
      Fastest thing Alive
    </h2>
    <img
      alt="sonic chars"
      src="https://b7.pngbarn.com/png/547/724/sonic-the-hedgehog-2-sonic-3d-sonic-mania-sonic-cd-sonic-the-hedgehog-png-clip-art.png"
    />
  </div>
</div>
`;
